Add contributing guide (#618)
open-telemetry/opentelemetry-proto: CONTRIBUTING.md

Welcome to the OpenTelemetry Protocol (OTLP) Specification repository! 🎉

Thank you for your interest in contributing to this project. This repository contains the OTLP protocol specification and the corresponding Language Independent Interface Types (.proto files). Your contributions help improve the interoperability of OpenTelemetry client libraries across various languages.

If you have any questions, feel free to ask in the community channels. We’re happy to help! 😊

## Prerequisites

Before getting started, ensure you have the following installed:

- **Docker** – [Install Docker](https://docs.docker.com/engine/install/)
- **Make** (for running development tasks)
- **Protobuf Compiler (protoc)** – [Install protoc](https://grpc.io/docs/protoc-installation/)

Additional Notes:
- Ensure your environment supports the required dependencies for building and testing .proto files.
- Windows users may need [Git Bash](https://gitforwindows.org/) for better compatibility.

## Workflow

We follow a structured workflow to ensure smooth collaboration:

### Branch Naming Convention
- **Feature branches**: `feature/<short-description>`
- **Bugfix branches**: `fix/<short-description>`

> - **Documentation updates**: `docs/<short-description>`

Reviewer: Do we follow this? Also, branches are all on forks. So the branch name doesn't really matter.

**Error:** `protoc: command not found`

**Fix:** Install the protobuf compiler:
```bash
sudo apt install protobuf-compiler  # Linux
brew install protobuf  # macOS
```

#### 2. Build Failures
**Error:** `Failed to generate gRPC client libraries`

**Fix:** Ensure Docker is running and retry:
```bash
make clean && make gen-all
```

#### 3. Linting Errors
**Error:** `proto-lint: some files do not conform to style guidelines`

**Fix:** Run:
```bash
make proto-lint-fix
```

---

## Additional Information

### Stability Guidelines

This repository follows OpenTelemetry’s stability guarantees:
- Stable components will not introduce breaking changes.
- Development components may be removed or changed without prior notice.

Refer to the [Versioning and Stability](https://github.com/open-telemetry/opentelemetry-proto/blob/main/README.md#versioning-and-stability) documentation for more details.

### Generating gRPC Client Libraries

To generate raw gRPC client libraries for supported languages, run:

```bash
make gen-<language>
```

Currently supported languages:
- cpp
- csharp
- go
- java
- objc
- openapi (swagger)
- php
- python
- ruby

Thank you for contributing! 🚀

> 

Reviewer: The remaining doc below seems to now be a duplicate of the rest of the document?

# Contributing to OpenTelemetry Protocol (OTLP) Specification

## Introduction

Welcome to the OpenTelemetry Protocol (OTLP) Specification repository! 🎉

> Thank you for your interest in contributing to this project. This repository contains the OTLP specification and the corresponding Language Independent Interface Types (`.proto` files). Your contributions help improve the interoperability of OpenTelemetry clients across various programming languages.

Reviewer 2: Why repeating what is already in `README.md`?

- Link relevant issues in the PR description.

---

## Local Setup & Development

To set up your local development environment:

```bash
# Clone the repository
git clone https://github.com/open-telemetry/opentelemetry-proto.git
cd opentelemetry-proto

# Generate all implementations
make gen-all
```

## Testing

To test your changes, run:

```bash
make test
```

To verify the protobuf definitions:

```bash
make proto-lint
```

## Contributing Rules

- Follow OpenTelemetry’s [Contribution Guidelines](https://github.com/open-telemetry/community/blob/main/guides/contributor/README.md).
- Ensure that `.proto` files adhere to OpenTelemetry's stability guarantees.
- Include clear and concise documentation updates if needed.

Check for issues labeled [`good first issue`](https://github.com/open-telemetry/opentelemetry-proto/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) to start contributing.

## Further Help

Need help? Join our community:

- **Slack**: [OpenTelemetry Slack](https://opentelemetry.io/community/)
- **Issues**: If you encounter a bug, [open an issue](https://github.com/open-telemetry/opentelemetry-proto/issues)

## Troubleshooting

> #### Protobuf Compiler Issues

Reviewer 2: You should decrease the header level

# Contributing to OpenTelemetry Protocol (OTLP) Specification

## Introduction

Welcome to the OpenTelemetry Protocol (OTLP) Specification repository! 🎉

> Thank you for your interest in contributing to this project. This repository contains the OTLP protocol specification and the corresponding Language Independent Interface Types (.proto files). Your contributions help improve the interoperability of OpenTelemetry client libraries across various languages.

Reviewer 3: ```suggestion
Thank you for your interest in contributing to this project. This repository contains the OTLP specification and the corresponding Language Independent Interface Types (`.proto` files). Your contributions help improve the interoperability of OpenTelemetry clients across various programming languages.
```

> If you have any questions, feel free to ask in the community channels. We’re happy to help! 😊

Reviewer 3: ```suggestion
If you have any questions, feel free to ask in the community channels. We're happy to help! 😊
```

## Prerequisites

Before getting started, ensure you have the following installed:

- **Docker** – [Install Docker](https://docs.docker.com/engine/install/)
- **Make** (for running development tasks)
- **Protobuf Compiler (protoc)** – [Install protoc](https://grpc.io/docs/protoc-installation/)

Additional Notes:
- Ensure your environment supports the required dependencies for building and testing .proto files.
- Windows users may need [Git Bash](https://gitforwindows.org/) for better compatibility.

> 
> 

Reviewer 3: ```suggestion

```

## Workflow

We follow a structured workflow to ensure smooth collaboration:

> 
> 

Reviewer 3: ```suggestion

```

> ### Commit Message Format
> - Use descriptive commit messages (e.g., `fix(proto): resolve issue with trace spec`)

Reviewer 3: ```suggestion
### Commit Message Format

- Use descriptive commit messages (e.g., `fix(proto): resolve issue with trace spec`)
```

- Follow [Conventional Commits](https://www.conventionalcommits.org/) where possible.

> ### Pull Request Guidelines
> - Fork the repository and create a new branch.

Reviewer 3: ```suggestion
### Pull Request Guidelines

- Fork the repository and create a new branch.
```

Reviewer 3: There are more places that share the same issue, I didn't comment on all of them (to reduce noise). @RichardChukwu would you go through the doc and fix them all? Looks like we need markdownlint rules.

Author: Yeah, just went over the mardown link errors now too and fixing broken links now

- Follow the coding guidelines before submitting your PR.
- Ensure tests pass locally before pushing.
- Link relevant issues in the PR description.

---

## Local Setup & Development

To set up your local development environment:

```bash
# Clone the repository
git clone https://github.com/open-telemetry/opentelemetry-proto.git
cd opentelemetry-proto

# Generate all implementations
make gen-all
```

To validate your changes:
```bash
make check
```

To update dependencies:
```bash
make update
```

---

## Testing

To test your changes, run:

```bash
make test
```

To verify the protobuf definitions:
```bash
make proto-lint
```

To check compatibility with different OpenTelemetry client libraries:
```bash
make compatibility-test
```

---

## Contributing Rules

- Follow OpenTelemetry’s [Contribution Guidelines](https://github.com/open-telemetry/community/blob/main/CONTRIBUTING.md).

> - Ensure that .proto files adhere to OpenTelemetry’s stability guarantees.

Reviewer 3: ```suggestion
- Ensure that `.proto` files adhere to OpenTelemetry's stability guarantees.
```

Reviewer 3: Can we ensure this via CI job?

Author: Hmmmm... is there a pipeline for that already via Github acitons?

> - Ensure new changes maintain compatibility with existing OpenTelemetry client libraries.

Reviewer 3: Could we explain a bit (maybe give an example) regarding how the contributors can do this? (e.g. we certainly don't expect contributors to test all the client libraries).

Reviewer 3: If we can make it clear and actionable, it will be great.
If it takes time to make it clear and actionable, I'm fine removing it for now rather than keeping something that is hard to understand and execute.

Author: Ok, let's remove for now then

# Contributing to OpenTelemetry Protocol (OTLP) Specification

## Introduction

Welcome to the OpenTelemetry Protocol (OTLP) Specification repository! 🎉

Thank you for your interest in contributing to this project. This repository contains the OTLP specification and the corresponding Language Independent Interface Types (`.proto` files). Your contributions help improve the interoperability of OpenTelemetry clients across various programming languages.

If you have any questions, feel free to ask in the community channels. We're happy to help! 😊

## Prerequisites

Before getting started, ensure you have the following installed:

* **Docker** – [Install Docker](https://docs.docker.com/engine/install/)
* **Make** (for running development tasks)
* **Protobuf Compiler (protoc)** – [Install protoc](https://grpc.io/docs/protoc-installation/)

Additional Notes:

> * Ensure your environment supports the required dependencies for building and testing .proto files.

Reviewer 4: Such as?

* Windows users may need [Git Bash](https://gitforwindows.org/) for better compatibility.

## Workflow

We follow a structured workflow to ensure smooth collaboration:

### Commit Message Format

* Use descriptive commit messages (e.g., `fix(proto): resolve issue with trace spec`)
* Follow [Conventional Commits](https://www.conventionalcommits.org/) where possible.

### Pull Request Guidelines

- Fork the repository and create a new branch.

> - Follow the coding guidelines before submitting your PR.

Reviewer 2: What are the coding guidelines?

> - Ensure tests pass locally before pushing.

Reviewer 4: ```suggestion
- Ensure `make all` pass locally before pushing.
```

- Link relevant issues in the PR description.

---

## Local Setup & Development

To set up your local development environment:

```bash
# Clone the repository
git clone https://github.com/open-telemetry/opentelemetry-proto.git
cd opentelemetry-proto

# Generate all implementations
make gen-all
```

To validate your changes:

```bash

> make check

Reviewer 4: There is no `check` target as far as I can tell in the makefile.

```

To update dependencies:

```bash

> make update

Reviewer 4: Dito, there is no `update` target. Are instructions copy/pasted from somewhere else? Please remove irrelevant parts.

```

## Testing

To test your changes, run:

```bash
make test
```

To verify the protobuf definitions:

```bash
make proto-lint
```

To check compatibility with different OpenTelemetry client libraries:

```bash
make compatibility-test
```

## Contributing Rules

> - Follow OpenTelemetry’s [Contribution Guidelines](https://opentelemetry.io/docs/contributing/style-guide/).

Reviewer 2: Is this not for https://github.com/open-telemetry/opentelemetry.io/?

> - Ensure that `.proto` files adhere to OpenTelemetry's stability guarantees.

Reviewer 2: Maybe a hyperlink?

- Include clear and concise documentation updates if needed.

Check for issues labeled [`good first issue`](https://github.com/open-telemetry/opentelemetry-proto/issues?q=is%3Aissue+is%3Aopen+label%3A%22good+first+issue%22) to start contributing.

## Further Help

Need help? Join our community:

- **Slack**: [OpenTelemetry Slack](https://opentelemetry.io/community/)
- **Issues**: If you encounter a bug, [open an issue](https://github.com/open-telemetry/opentelemetry-proto/issues)

> ## Troubleshooting Guide
> 
> ### Common Issues & Fixes

Reviewer 2: I think this can be simplified 

```suggestion
## Troubleshooting
```

> #### 1. Protobuf Compiler Issues

Reviewer 2: Why the numbers?

**Error:** `protoc: command not found`

**Fix:** Install the protobuf compiler:
```bash
sudo apt install protobuf-compiler  # Linux
brew install protobuf  # macOS
```

#### 2. Build Failures

**Error:** `Failed to generate gRPC client libraries`

**Fix:** Ensure Docker is running and retry:
```bash
make clean && make gen-all
```

#### 3. Linting Errors

**Error:** `proto-lint: some files do not conform to style guidelines`

**Fix:** Run:
```bash
make proto-lint-fix
```

---

## Additional Information

### Stability Guidelines

This repository follows OpenTelemetry’s stability guarantees:

* Stable components will not introduce breaking changes
* Development components may be removed or changed without prior notice.

> Refer to the [Versioning and Stability](https://github.com/open-telemetry/opentelemetry-proto/blob/main/README.md#versioning-and-stability) documentation for more details.

Reviewer 2: Use relative hyperlink
